cirq-core/cirq/optimizers/two_qubit_to_sqrt_iswap.py

```diff
@@ -0,0 +1,340 @@
+        Towards ultra-high fidelity quantum operations: SQiSW gate as a native
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    kak = linalg.kak_decomposition(
+        mat, atol=atol / 10, rtol=0, check_preconditions=check_preconditions
+    )
+    operations = _kak_decomposition_to_sqrt_iswap_operations(
+        q0, q1, kak, required_sqrt_iswap_count, include_global_phase=not clean_operations, atol=atol
+    )
+    if clean_operations:
+        return two_qubit_decompositions._cleanup_operations(operations)
+    return operations
+
+
+def _kak_decomposition_to_sqrt_iswap_operations(
+    q0: 'cirq.Qid',
+    q1: 'cirq.Qid',
+    kak: linalg.KakDecomposition,
+    required_sqrt_iswap_count: Optional[int] = None,
+    include_global_phase: bool = False,
+    atol: float = 1e-8,
+) -> Sequence['cirq.Operation']:
+    """Computes the list of operations in the sqrt-iSWAP decomposition."""
+    single_qubit_operations, global_phase = _single_qubit_matrices_with_sqrt_iswap(
+        kak, required_sqrt_iswap_count, atol=atol
+    )
+    return _decomp_to_operations(
+        q0,
+        q1,
+        ops.SQRT_ISWAP,
+        single_qubit_operations,
+        global_phase if include_global_phase else None,
+        atol=atol,
+    )
+
+
+def _decomp_to_operations(
+    q0: 'cirq.Qid',
+    q1: 'cirq.Qid',
+    two_qubit_gate: 'cirq.Gate',
+    single_qubit_operations: Sequence[Tuple[np.ndarray, np.ndarray]],
+    global_phase: Optional[complex] = None,
+    atol: float = 1e-8,
+) -> Sequence['cirq.Operation']:
+    """Converts a sequence of single-qubit unitary matrices on two qubits into a
+    list of operations with interleaved two-qubit gates."""
+    two_qubit_op = two_qubit_gate(q0, q1)
+    operations = []
+    if global_phase is not None:
+        operations.append(ops.GlobalPhaseOperation(global_phase, atol))
+
+    def append(matrix0, matrix1):
+        operations.append(ops.MatrixGate(matrix0).on(q0))
+        operations.append(ops.MatrixGate(matrix1).on(q1))
+
+    iter_ops = iter(single_qubit_operations)
+    for matrix0, matrix1 in iter_ops:
+        append(matrix0, matrix1)  # Append first pair of single qubit gates
+        break
+    for matrix0, matrix1 in iter_ops:
+        operations.append(two_qubit_op)  # Append two-qubit gate between each pair
+        append(matrix0, matrix1)  # Append other pairs of single qubit gates
+    return operations
+
+
+def _single_qubit_matrices_with_sqrt_iswap(
+    kak: 'cirq.KakDecomposition',
+    required_sqrt_iswap_count: Optional[int] = None,
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Computes the sequence of interleaved single-qubit unitary matrices in the
+    sqrt-iSWAP decomposition."""
+    if required_sqrt_iswap_count is not None:
+        if not 0 <= required_sqrt_iswap_count <= 3:
+            raise ValueError('the argument `required_sqrt_iswap_count` must be 0, 1, 2, or 3.')
+        if not [
```

Comment:
Wouldn't this fail if I try to decompose Identity / CNOT etc. using 3 Sqrt ISwaps? `_in_3sqrt_iswap_region` always returns true, but it should return true only for region `x < y + |z|`, since we are checking for "exactly 3" and not "at most 3" ?

Reply:
*Any* unitary can be decomposed into exactly 3 sqrt-iSWAPs.

Reply:
`required_sqrt_iswap_count=3` would be useful for compiling parameterized circuits and always get the same size circuit (similar noise characteristics).